plugins/BEdita/Core/src/Model/Entity/Tree.php

```diff
@@ -1,38 +1,116 @@
  *
  * @property \BEdita\Core\Model\Entity\ObjectEntity $object
- * @property \BEdita\Core\Model\Entity\ParentTree $parent_tree
- * @property \BEdita\Core\Model\Entity\ChildTree[] $child_trees
+ * @property \BEdita\Core\Model\Entity\ObjectEntity $parent_object
+ * @property \BEdita\Core\Model\Entity\ObjectEntity $root_object
+ * @property \BEdita\Core\Model\Entity\Tree $parent_node
+ * @property \BEdita\Core\Model\Entity\Tree[] $child_nodes
+ *
+ * @since 4.0.0
  */
 class Tree extends Entity
 {
 
     /**
-     * Fields that can be mass assigned using newEntity() or patchEntity().
-     *
-     * Note that when '*' is set to true, this allows all unspecified fields to
-     * be mass assigned. For security purposes, it is advised to set '*' to false
-     * (or remove it), and explicitly make individual fields accessible as needed.
-     *
-     * @var array
+     * {@inheritDoc}
      */
     protected $_accessible = [
-        '*' => true,
-        'id' => false
+        '*' => false,
+        'object_id' => true,
+        'parent_id' => true,
+        'menu' => true,
     ];
+
+    /**
+     * Setter for `parent_id`.
+     * It also set `root_id` and `parent_node_id`.
+     *
+     * When `$parent_id` is null it means that the node is a root (first level on tree) so
+     * - the `root_id` is set equal to `object_id`
+     * - the parent_node_id is set to `null`
+     *
+     * When `$parent_id` is not null the parent node is retrieved and
+     * - the `root_id` is set equal to parent node `root_id`
+     * - the `parent_node_id` is set to the parent node `id`
+     *
+     * @param int|null $parentId The value to set
+     * @return int|null
+     */
+    protected function _setParentId($parentId)
+    {
+        if (array_key_exists('parent_id', $this->_properties) && $this->_properties['parent_id'] === $parentId) {
+            return $parentId;
+        }
+
+        if ($parentId === null) {
+            $this->root_id = $this->object_id;
+            $this->parent_node_id = null;
+
+            return $parentId;
+        }
+
+        // set root_id and parent_node_id
+        $table = TableRegistry::get($this->getSource());
+        $parentNode = $table
+            ->find()
+            ->where(['object_id' => $parentId])
+            ->firstOrFail();
+
+        $this->root_id = $parentNode->root_id;
+        $this->parent_node_id = $parentNode->id;
+
+        return $parentId;
+    }
+
+    /**
+     * Setter for `parent_object` property.
+     *
+     * Set the related `parent_id` if changed.
+     * In this way `parent_node_id` will be marked as dirty (see `self::_setParentId()`)
+     * ensuring the recover of the tree nested set model.
+     *
+     * @param \BEdita\Core\Model\Entity\Folder|null $folder The folder entity to set as parent
+     * @return \BEdita\Core\Model\Entity\Folder|null
+     */
+    protected function _setParentObject(Folder $folder = null)
+    {
+        $parentId = ($folder === null) ? null : $folder->id;
```

Comment:
I'm not a super fan of ternary operators… they make line coverage less relevant. I'd prefer:

```php
$parentId = null;
if ($folder !== null) {
    $parentId = $folder->id;
}
```